### PR TITLE
Debian package for Barefoot Tofino-based platform

### PR DESCRIPTION
**Pull request by Yi-Tseng**

### CODE REVIEW

File: stratum/hal/bin/barefoot/BUILD [hunk 1]
```diff
@@ -56,3 +57,98 @@ stratum_cc_binary(
...
+    name = "stratum_bf_deb",
+    architecture = "amd64",
+    data = ":stratum_bf_data",
+    depends = [
+        "systemd"
+    ],
+    description_file = "deb/description",
+    homepage = "https://stratumproject.org/",
+    maintainer = "https://stratumproject.org/",
+    package = "stratum_bf",
```
Comment (pudelkoM) [inline stratum/hal/bin/barefoot/BUILD:151]: Debian uses dashes in package names

File: stratum/hal/bin/barefoot/BUILD [hunk 2]
```diff
@@ -56,3 +57,98 @@ stratum_cc_binary(
...
 )
+
+# Rules for Debian package
+STRIP_PREFIX="/stratum/hal/bin/barefoot"
+
+pkg_tar(
+    name = "stratum_bf_bin",
+    srcs = [":stratum_bf"],
+    mode = "0755",
+    package_dir = "/usr/local/bin",
```
Comment (Yi-Tseng) [inline stratum/hal/bin/barefoot/BUILD:68]: move things to `/usr` instead of `/usr/local`

**Comment by Yi-Tseng** [comment]
WIP:
 - huge page commands in entry point script
 - remove pid file stuff
 - use `exec` to run stratum in entry point script

File: stratum/hal/bin/barefoot/BUILD [hunk 3]
```diff
@@ -56,3 +57,107 @@ stratum_cc_binary(
...
+    strip_prefix = STRIP_PREFIX,
+)
+
+pkg_tar(
+    name = "stratum_entry_point",
+    srcs = ["deb/stratum-entrypoint"],
+    mode = "0755",
+    package_dir = "/usr/bin",
+    strip_prefix = STRIP_PREFIX + "/deb",
+)
```
Comment (pudelkoM) [inline stratum/hal/bin/barefoot/BUILD:78]: I think these two rules can be combined. By default `pkg_tar` will flatten the directory hierarchy.

File: stratum/hal/bin/barefoot/README.md [hunk 4]
```diff
@@ -223,3 +223,21 @@ To start a shell session, you can use (requires Docker):
...
+
+The Debian package includes some necessary libraries like PI so you don't need to install these libraries into the target device.
+However, you still need to install some dependency packages such as **libboost**, **kmod**, and so on. Below are dependencies that required by this package:
+
+ - systemd
+ - kmod
+ - libssl1.1
+ - libedit2
+ - libjudydebian1
+ - libboost-thread1.62.0
```
Comment (pudelkoM) [inline stratum/hal/bin/barefoot/README.md:243]: Is that still true? I see them listed in `depends`

Comment (Yi-Tseng) [reply stratum/hal/bin/barefoot/README.md:243]: I think we should add command here which tells the user how to install the package.
as far as I know there are two ways, one is to use `apt install`, another is to use `dpkg -i`
`apt install` will try to install packages automatically
`dpkg -i` command will install all files to the system, but will show errors below:
```bash
root@d88707268ae6:/stratum# dpkg -i stratum_bf_deb.deb
Selecting previously unselected package stratum-bf.
(Reading database ... 6500 files and directories currently installed.)
Preparing to unpack stratum_bf_deb.deb ...
Unpacking stratum-bf (0.0.1) ...
dpkg: dependency problems prevent configuration of stratum-bf:
 stratum-bf depends on systemd; however:
  Package systemd is not installed.
 stratum-bf depends on kmod; however:
  Package kmod is not installed.
 stratum-bf depends on libssl1.1; however:
  Package libssl1.1 is not installed.
 stratum-bf depends on libedit2; however:
  Package libedit2 is not installed.
 stratum-bf depends on libjudydebian1; however:
  Package libjudydebian1 is not installed.
 stratum-bf depends on libboost-thread1.62.0; however:
  Package libboost-thread1.62.0 is not installed.

dpkg: error processing package stratum-bf (--install):
 dependency problems - leaving unconfigured
Errors were encountered while processing:
 stratum-bf
root@d88707268ae6:/stratum# echo $?
1
root@d88707268ae6:/stratum#
```

Comment (pudelkoM) [reply stratum/hal/bin/barefoot/README.md]: Yes, `dpkg` just unpacks files. You have to run `apt-get install -f` afterwards to install the dependencies. Or just use `apt-get` in the first place. My point is, that this list will get outdated and only encourages people to manually install Stratum.

File: stratum/hal/bin/barefoot/README.md [hunk 5]
```diff
@@ -223,3 +258,39 @@ To start a shell session, you can use (requires Docker):
...
+You can find the Debian package here:
+`bazel-bin/stratum/hal/bin/barefoot/stratum_bf_deb.deb`
+
+The Debian package includes some necessary libraries like PI so you don't need to install these libraries into the target device.
+
+To install the package, simply upload the deb file to your device and use `apt` command:
+
+```bash
+[sudo] apt update
+[sudo] apt install -y stratum_bf_deb.deb
```
Comment (pudelkoM) [inline stratum/hal/bin/barefoot/README.md:285]: Isnt't a `./` missing here?

I get this error without it:
```
apt-get install -y stratum_bcm_opennsa_deb.deb
Reading package lists... Done
Building dependency tree
Reading state information... Done
E: Unable to locate package stratum_bcm_opennsa_deb.deb
E: Couldn't find any package by glob 'stratum_bcm_opennsa_deb.deb'
E: Couldn't find any package by regex 'stratum_bcm_opennsa_deb.deb'
```

File: stratum/hal/bin/barefoot/README.md (hunk 4 above)
Comment (pudelkoM) [reply stratum/hal/bin/barefoot/README.md]: I think you found a good trade off here.

File: stratum/hal/bin/barefoot/BUILD [hunk 6]
```diff
@@ -56,3 +57,78 @@ stratum_cc_binary(
...
+    data = ":stratum_bf_data",
+    depends = [
+        "systemd",
+        "kmod",
+        "libssl1.1",
+        "libedit2",
+        "libjudydebian1",
+        "libboost-thread1.62.0",
+    ],
+    description_file = "deb/description",
```
Comment (pudelkoM) [inline stratum/hal/bin/barefoot/BUILD:127]: What's the benefit of `description_file` over the inline `description`?

File: stratum/hal/bin/barefoot/BUILD [hunk 7]
```diff
@@ -56,3 +57,78 @@ stratum_cc_binary(
...
+    homepage = "https://stratumproject.org/",
+    maintainer = "https://stratumproject.org/",
```
Comment (pudelkoM) [inline stratum/hal/bin/barefoot/BUILD:129]: maintainer = "The Stratum Project"

File: stratum/hal/bin/barefoot/docker/Dockerfile.runtime [hunk 8]
```diff
@@ -35,29 +35,14 @@ ENV KERNEL_HEADERS_PATH=/usr/src/kernel-headers
...
-    touch $OUTPUT_BASE/share/stratum/phal_config.pb.txt && \
-    cp $STRATUM_BF_BASE/chassis_config.pb.txt $OUTPUT_BASE/share/stratum/ && \
-    mkdir -p /output/stratum/hal/lib/common/ && \
-    cp stratum/hal/lib/common/gnmi_caps.pb.txt /output/stratum/hal/lib/common/
-    # TODO(Yi Tseng): Remove two lines above when we are able to generate cap response automatically
+# Stratum package
+RUN bazel build //stratum/hal/bin/barefoot:stratum_bf_deb \
+    --define sde_ver=$(cat $SDE_INSTALL/share/VERSION) \
+    --define phal_with_onlp=$WITH_ONLP
+RUN dpkg -x bazel-bin/stratum/hal/bin/barefoot/stratum-bf_0.0.1_amd64.deb $OUTPUT_BASE
```
Comment (pudelkoM) [inline stratum/hal/bin/barefoot/docker/Dockerfile.runtime:45]: I think there is also a deb file without the version. So we don't have to change this later.

**Pull request merged**